Grab Bag Loot Modernization

BasicallyIAmFox opened the pull request:
### What is the new feature?
This new feature replaces OpenVanillaBag, PreOpenVanillaBag and few other methods with new ModifyItemLoot and ModifyGlobalLoot in correspondent classes.
It could allow for mods such as Boss Checklist to check boss bag loot without need for other mod to add all loot for its boss manually in Boss Checklist.

### Why should this be part of tModLoader?
I feel like this might be difficult if you want to add new item to vanilla pool, and NPC Loot way allows you to do that (even if its not best way read-wise), and so I had to IL edit vanilla [insert literally any method that can spawn loot] method to add just single item in the item pool.

### Are there alternative designs?
No, I don't think so?

### Sample usage for the new feature
Same as ModifyNPCLoot pretty much.

BasicallyIAmFox committed:
lots of changes

TheLuxure commented:
TODO/REDO: Herb Bags, Presents and Goodie Bags dont work (needs fix)
After finishing: Example Mod changes

BasicallyIAmFox committed:
Update ItemDropDatabase.cs.patch

TheLuxure commented:
![image](https://user-images.githubusercontent.com/65787578/179334011-f13e2cc9-d17a-439e-9cb0-7d4c360b9c34.png)
kinda offtopic, but because of new way i managed to make tha thting in terraria

BasicallyIAmFox committed:
Update ItemDropDatabase.cs.patch

BasicallyIAmFox committed:
Update ItemDropDatabase.cs.patch

BasicallyIAmFox committed:
the return of CursorOverrideID

BasicallyIAmFox committed:
fuixes

TheLuxure commented:
should i add new ItemDropRules, such as with ability to drop 2 items from options or more?

BasicallyIAmFox committed:
meow

BasicallyIAmFox committed:
gh

BasicallyIAmFox committed:
screaming for help

ExterminatorX99 committed:
Cleanup patches and code

ExterminatorX99 committed:
Cleanup `FewFrom` drop rules

ExterminatorX99 committed:
More patch context, remove using

BasicallyIAmFox committed:
quite a few changes/fixes

fixed Goodie Bag and Presents (huge thanks to Snek)
added new droprules (SequentialRulesRule and SequentialRulesNotScalingWithLuckRule correspondently, both by Snek)
added global entries for boss bags, presents, herb bags, and crates. (not forcing users to use ModifyItemLoot instead anymore)

fixed ExampleFishingCrate not able to drop baits.

BasicallyIAmFox committed:
hopefully fixed dev sets

BasicallyIAmFox committed:
oyster and can of worms

BasicallyIAmFox committed:
bramble crate fix

BasicallyIAmFox committed:
feeeeeding back

BasicallyIAmFox committed:
forgor

ExterminatorX99 committed:
Merge remote-tracking branch 'TheLuxure/1.4-existingvmethodstonpclootwayliek' into 1.4-existingvmethodstonpclootwayliek

# Conflicts:
#	ExampleMod/Content/Items/Consumables/ExampleFishingCrate.cs
#	patches/tModLoader/Terraria/GameContent/ItemDropRules/FewFromOptionsDropRule.cs
#	patches/tModLoader/Terraria/GameContent/ItemDropRules/FewFromOptionsNotScaledWithLuckDropRule.cs
#	patches/tModLoader/Terraria/Player.cs.patch

BasicallyIAmFox committed:
Merge pull request #1 from ExterminatorX99/1.4-existingvmethodstonpclootwayliek

Cleanup

BasicallyIAmFox committed:
few food

BasicallyIAmFox committed:
herb bag fix

BasicallyIAmFox committed:
patch

BasicallyIAmFox committed:
e

BasicallyIAmFox committed:
fixed crash

BasicallyIAmFox committed:
commented orig code

BasicallyIAmFox committed:
fixed FewFromXDropRule dropping + 1 pieces

Chicken-Bones committed:
Remove CursorOverrideID patch

Chicken-Bones committed:
Restore old hooks with obsolete modifiers.

Chicken-Bones committed:
A bit of patch/styling/code reuse cleanup

Chicken-Bones committed:
Cleanup ItemDropRules.CommonCode
